feat(ci): add openapi spec sync check (#1721)
namehash/ensnode: docs/ensnode.io/src/content/docs/docs/contributing/rest-api-docs.mdx

---
title: REST API Documentation
description: How REST API documentation is generated and kept in sync.
sidebar:
  label: REST API Docs
  order: 5
---

> The REST API docs are generated by Mintlify from an OpenAPI spec. The source of that OpenAPI spec depends on the context of a doc build. This means the [production REST API docs](https://docs.ensnode.io/ensapi) always reflect the version of the ENSApi instance actively running in production (often different than what's currently committed to the main branch), while preview REST API docs reflect upcoming REST API changes as committed to code in a PR (independent of any deployment of ENSApi).

Reviewer: ```suggestion
Our REST API docs are generated by Mintlify from an OpenAPI spec. The source of that OpenAPI spec depends on the context of a Mintlify Docs build. XXX, while preview Mintlify REST API docs reflect upcoming REST API changes as committed to code in a PR (independent of any deployment of ENSApi).
```

  order: 5
---

> Our REST API docs are generated by Mintlify from an OpenAPI spec. The Mintlify site is configured (in [`docs.json`](https://github.com/namehash/ensnode/blob/main/docs/docs.ensnode.io/docs.json)) with two API doc groups that use different OpenAPI spec sources:

Reviewer: ```suggestion
Our REST API docs are generated by Mintlify from an OpenAPI spec. The Mintlify site is configured (in [`docs/docs.ensnode.io/docs.json`](https://github.com/namehash/ensnode/blob/main/docs/docs.ensnode.io/docs.json)) with two API doc groups that use different OpenAPI spec sources:
```

| Group             | Visibility | OpenAPI Source                                                                                                                 | Purpose                                               |
| ----------------- | ---------- | ------------------------------------------------------------------------------------------------------------------------------ | ----------------------------------------------------- |
| **API Reference** | Visible    | [Live production ENSApi instance](https://api.alpha.ensnode.io/openapi.json)                                                   | Always reflects the currently deployed production API |

> | **Preview**       | Hidden     | Committed [`ensapi-openapi.json`](https://github.com/namehash/ensnode/blob/main/docs/docs.ensnode.io/ensapi-openapi.json) file | Shows upcoming API changes from the source branch     |

Reviewer: ```suggestion
| **Preview**       | Hidden     | Committed [`ensapi-openapi.json`](https://github.com/namehash/ensnode/blob/main/docs/docs.ensnode.io/ensapi-openapi.json) file | Shows REST API docs based on the source branch     |
```

  order: 5
---

The [REST API Docs](https://docs.ensnode.io/ensapi) are generated by Mintlify from an OpenAPI spec. The source of that spec depends on the deployment context. This means production docs always reflect the version of the API running in production, while PR previews document upcoming API changes before they're deployed to production.

### Production Deployments

> When docs are deployed to production (on merge to `main`), Mintlify fetches the OpenAPI spec directly from the live API.

Reviewer: ```suggestion
On each commit to the `main` branch, Mintlify automatically deploys a rebuild of the production docs as follows:
```

> | Content Type       | Source                                                                                                                                            | Output                                                     |

Reviewer: ```suggestion
| Content       | Source                                                                                                                                            | Output                                                     |
```

| ------------------ | ------------------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------- |
| **REST API Docs**  | OpenAPI spec fetched from `https://api.alpha.ensnode.io/openapi.json` at build time (published by the ENSApi service running in production) | [Production REST API Docs](https://docs.ensnode.io/ensapi) |
| **All other docs** | Committed files on `main`                                                                                                                         | Deploys immediately on merge to `main`                     |

Note: the production OpenAPI spec is generally different than the committed [`docs/docs.ensnode.io/ensapi-openapi.json`](https://github.com/namehash/ensnode/blob/main/docs/docs.ensnode.io/ensapi-openapi.json) file, since the committed file may contain upcoming changes not yet deployed to the API. Non-API documentation (guides, concepts, etc.) may also be published before the corresponding code is released to production.

> ### Preview Deployments

Reviewer: What about generating "Preview Doc Builds" locally? Is there an easy process for this? If so, it will be great to document that here. If this isn't possible and all "Preview Doc Builds" should be accessed via PRs, we should explicitly document that here.

> Mintlify creates a deployment preview for every pull request that changes files in the docs directory. The preview URL is posted as a comment on the PR. In preview deployments, the REST API Docs are generated from the committed OpenAPI spec file rather than the live API, so you can review upcoming API changes before they're deployed.

Reviewer: Can you please make it more explicit here exactly what the "docs directory" is? In my mind this could be interpreted in multiple ways and I'm not sure which is correct.
